kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ProduceRequest.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.kafkarest.entities.v3;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.BinaryNode;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.ByteString;
+import io.confluent.kafka.schemaregistry.ParsedSchema;
+import io.confluent.kafka.serializers.subject.RecordNameStrategy;
+import io.confluent.kafka.serializers.subject.TopicNameStrategy;
+import io.confluent.kafka.serializers.subject.TopicRecordNameStrategy;
+import io.confluent.kafka.serializers.subject.strategy.SubjectNameStrategy;
+import io.confluent.kafkarest.entities.EmbeddedFormat;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+@AutoValue
+public abstract class ProduceRequest {
+
+  ProduceRequest() {
+  }
+
+  @JsonProperty("partition_id")
+  @JsonInclude(Include.NON_ABSENT)
+  public abstract Optional<Integer> getPartitionId();
```

Comment:
my instinct is this would be better specified in the url. i.e.:

`/v3/clusters/{clusterId}/topics/{topicName}/partitions/{partitionId}/records`

and not be specifiable in the JSON when producing at the topic level.

Reply:
I'm receiving partitionId in the body to allow production to multiple partitions using a single streaming connection. That would be the behaviour of the Java API. I do agree that we should add an specialization of the API that receives the partitionId in the URL. I will defer it to a future PR.